### PR TITLE
feat: adding PP and CP for nemotron v3 models

### PR DESCRIPTION
**Pull request by adil-a**
# What does this PR do ?

Adds **pipeline parallelism (PP)** and **context parallelism (CP)** support for the NemotronV3/H (hybrid mamba + attention + MoE) model family, including Multi-Token Prediction (MTP) and THD sequence packing, plus the supporting MoE checkpoint and shared-expert changes these models need.

# Changelog

**Pipeline parallelism (NemotronV3/H)**
- Per-stage forward contract: mid stages return `(hidden_states, *mtp_embed_inputs)`; the last stage returns `(logits, *mtp_per_depth_h, seq_idx)`. The trailing `[B, S] int32` `seq_idx` binds each microbatch's sub-sequence layout to its own loss call via the PP runtime's output→loss contract, so the wiring is schedule-agnostic.
- MTP embed-input propagation between stages and last-stage-only MTP head ownership.

**Context parallelism (NemotronV3/H)**
- CP for both attention and the Mamba2 mixer (TE p2p / DualChunkSwap and DTensor `context_parallel()` + SDPA), with THD-format CP batch sharding.

**Multi-Token Prediction (MTP)**
- MTP head with optional repeated-layer reuse; per-depth label rolling with trailing-`k+1` masking.
- `PipelineCausalLMLoss` adds the MTP auxiliary CE on the last stage; supports both hidden-state and pre-computed-logits inputs.
- Cross-sub-sequence label-roll masking via `seq_idx` (preferred) with a `cu_seqlens` fallback for models that don't emit a `seq_idx` tail.

**THD sequence packing (`thd_utils`)**
- Emit real `cu_seqlens` plus a separate `cu_seqlens_padded` and `max_seqlen` that honor Transformer Engine's THD contract; trailing-pad absorption for the single-pack case.
- Mamba `seq_idx` derived from the padded layout (`searchsorted(..., right=True)`) so the SSD scan segments packed bins correctly at micro-batch > 1.
- Gate the `cu_seqlens`-from-2D-`attention_mask` recovery on batch size == 1 (a single flattened stream); at B>1 each row is an independent sequence, so cu_seqlens is left unset and each consumer treats every row on its own.

**MoE supporting changes**
- Backward-correct shared-expert stream overlap: `_SharedExpertStreamFork` / `_SharedExpertStreamJoin` identity autograd functions add bidirectional stream fencing + `record_stream` so the side-stream overlap is correct in both forward and backward.
- Checkpoint `to_hf` expert splitting uses in-place strided views into the model's grouped storage (avoids large per-expert scratch on memory-tight nodes), with a save-time materialization helper for safetensors.

**Recipe (`train_ft`)**
- Gate THD/cu_seqlens processing on the dataset being THD-packed rather than on TE attention being present on a rank, so PP stages without attention layers still build cu_seqlens downstream; plumb `cu_seqlens` into the non-PP MTP loss.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests? (unit + functional: PP/MTP, THD utils, `seq_idx`-from-`cu_seqlens`, MTP cross-boundary masking, MTP rolling, MoE state-dict incl. LoRA and in-place views, and context-parallel for attention / mamba / hybrid)
- [ ] Did you add or update any necessary documentation?

# Additional Information

- Enables PP and CP for the NemotronV3/H model family. CI (incl. the L0 unit suites and the L2 context-parallel functional tests) is green.

### CODE REVIEW

File: nemo_automodel/components/models/nemotron_v3/model.py [hunk 1]
```diff
@@ -139,42 +141,85 @@ def forward(
...
+        # Non-THD-collater path doesn't emit cu_seqlens — recover it from a
+        # 2D indexed attention_mask so mamba's seq_idx derivation has input.
+        if (
+            "cu_seqlens" not in kwargs
+            and attention_mask is not None
+            and attention_mask.dim() == 2
+            and attention_mask.dtype != torch.bool
+        ):
+            seq_lens = attention_mask.sum(dim=-1).to(torch.int32)
+            kwargs["cu_seqlens"] = F.pad(seq_lens.cumsum(0).to(torch.int32), (1, 0))
```
[the remark this answers is not in the document]
Comment (adil-a) [reply nemo_automodel/components/models/nemotron_v3/model.py:196]: Fixed in cb7e9ab28. Rather than guard each consumer, I gated the `cu_seqlens`-from-2D-mask derivation itself on `B==1` (model.py): at B>1 the rows are independent sequences, so cu_seqlens is simply not synthesized and all three consumers (mamba, the PP `seq_idx` tail, MTP kwargs forwarding) treat each row as one sequence. `layers.py` was restructured so B>1 BSHD mamba still gets per-row `zeros(B,S)` independent of cu_seqlens — no behavior change there. Also updated 3 stale `test_nemotron_v3_pp_mtp.py` assertions that predated the final-stage `seq_idx` tail. Verified on 4xH100: `moe/` + `models/nemotron_v3/` unit suites = 640 passed, 0 failed.

File: nemo_automodel/components/models/common/utils.py [hunk 2]
```diff
@@ -190,7 +186,6 @@ class BackendConfig:
     dispatcher_num_sms: int = 20
     dispatcher_share_token_dispatcher: bool = True
     dispatcher_async_dispatch: bool = False
-    disable_shared_expert_overlap: bool = False
```
Comment (hemildesai) [inline nemo_automodel/components/models/common/utils.py:193]: Can we keep this?

**Review by hemildesai — APPROVED** [review]
